File: frontend/scripts/check-complexity.mjs
// Copyright (c) 2026 Mechemsi. All rights reserved.
// Licensed under AGPLv3. See LICENSE file.
// Commercial licensing: info@mechemsi.com

/**
 * Check cyclomatic complexity of TypeScript/TSX files using ESLint.
 *
 * Usage: node scripts/check-complexity.mjs <max_complexity> [file1.tsx file2.ts ...]
 *
 * max_complexity: Maximum allowed cyclomatic complexity (default: 15).
 * If files are provided, only those are checked. Otherwise checks all
 * files under src/.
 *
 * Exit codes:
 *   0 — no functions above the threshold
 *   1 — at least one function above the threshold
 */

import { execSync } from "node:child_process";
import { existsSync } from "node:fs";

const args = process.argv.slice(2);
const maxComplexity = parseInt(args[0], 10) || 15;
const files = args.slice(1).filter((f) => existsSync(f));

const targets = files.length > 0 ? files.join(" ") : "src/";

// Run ESLint with only the complexity rule enabled
const cmd = [
  "npx",
  "eslint",
  "--no-eslintrc",
  "-c",
  '{"rules":{"complexity":["warn",' + maxComplexity + "]}}",
  "--parser",
  "@typescript-eslint/parser",
  "--plugin",
  "@typescript-eslint",
  "--ext",
  ".ts,.tsx",
  "--format",
  "stylish",
  targets,
].join(" ");

try {
  const output = execSync(cmd, {
    encoding: "utf-8",
    stdio: ["pipe", "pipe", "pipe"],
  });
  // If no warnings, ESLint exits 0 and output is empty or just whitespace
  if (output.trim()) {
    console.log(output.trim());
  }
  console.log(
    `All functions are below complexity ${maxComplexity}. No issues found.`
  );
  process.exit(0);
} catch (err) {
  const output = (err.stdout || "") + (err.stderr || "");
  if (output.includes("complexity")) {
    console.log(`Functions exceeding complexity threshold ${maxComplexity}:\n`);
    console.log(output.trim());
    console.log(
      `\nReduce complexity by extracting helper functions or simplifying control flow.`
    );
    process.exit(1);
  }
  // ESLint may exit non-zero for parsing errors — still report
  if (output.trim()) {
    console.log(output.trim());
  }
  console.log(
    `All functions are below complexity ${maxComplexity}. No issues found.`
  );
  process.exit(0);
}

File: scripts/check_complexity.py
# Copyright (c) 2026 Mechemsi. All rights reserved.
# Licensed under AGPLv3. See LICENSE file.
# Commercial licensing: info@mechemsi.com
"""Check cyclomatic complexity of Python files using radon.

Usage: python scripts/check_complexity.py <min_grade> [file1.py file2.py ...]

min_grade: Minimum grade to report (A-F). E.g., "C" reports C, D, E, F.
If files are provided, only those are checked. Otherwise checks all
files under the current directory.

Exit codes:
  0 — no functions at or above the threshold
  1 — at least one function at or above the threshold
"""

import subprocess
import sys


GRADE_ORDER = ["A", "B", "C", "D", "E", "F"]


def main() -> int:
    if len(sys.argv) < 2:
        print("Usage: check_complexity.py <min_grade> [file1.py ...]")
        return 2

    min_grade = sys.argv[1].upper()
    if min_grade not in GRADE_ORDER:
        print(f"ERROR: invalid grade '{min_grade}'. Must be one of: {GRADE_ORDER}")
        return 2

    files = sys.argv[2:]

    # Build radon command
    cmd = ["python3", "-m", "radon", "cc", "--min", min_grade, "--show-complexity"]
    if files:
        cmd.extend(files)
    else:
        cmd.append(".")

    result = subprocess.run(cmd, capture_output=True, text=True)
    output = result.stdout.strip()

    if not output:
        print(f"All functions are below grade {min_grade}. No issues found.")
        return 0

    # Filter out empty lines and file headers with no results
    lines = output.splitlines()
    has_issues = False
    report_lines = []

    for line in lines:
        stripped = line.strip()
        if not stripped:
            continue
        # radon outputs file paths followed by indented function lines
        # Function lines contain the grade letter like " C (12)"
        report_lines.append(line)
        # Check if this line has a complexity grade at or above threshold
        for grade in GRADE_ORDER[GRADE_ORDER.index(min_grade) :]:
            if f" {grade} (" in line or f" {grade} " in line:
                has_issues = True
                break

    if not has_issues:
        print(f"All functions are below grade {min_grade}. No issues found.")
        return 0

    print(f"Functions at or above grade {min_grade}:\n")
    for line in report_lines:
        print(line)

    print(f"\nGrades: A (1-5) simple | B (6-10) well-structured | "
          f"C (11-15) slightly complex | D (16-20) more complex | "
          f"E (21-25) very complex | F (26+) unmaintainable")

    return 1


if __name__ == "__main__":
    sys.exit(main())
